servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/CancellableSet.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright © 2018 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import io.servicetalk.concurrent.Cancellable;
+
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+
+import static io.servicetalk.concurrent.internal.ThrowableUtils.catchUnexpected;
+import static io.servicetalk.utils.internal.PlatformDependent.throwException;
+import static java.util.Collections.newSetFromMap;
+import static java.util.concurrent.atomic.AtomicLongFieldUpdater.newUpdater;
+
+final class CancellableSet implements Cancellable {
+    private static final AtomicLongFieldUpdater<CancellableSet> counterUpdater =
+            newUpdater(CancellableSet.class, "counter");
+    private final Set<Cancellable> set;
+    private volatile long counter;
+
+    /**
+     * Create a new instance.
+     */
+    CancellableSet() {
+        this(8);
+    }
+
+    /**
+     * Create a new instance.
+     * @param initialSize The initial size of the internal {@link Set}.
+     */
+    CancellableSet(int initialSize) {
+        set = newSetFromMap(new ConcurrentHashMap<>(initialSize));
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Cancel all {@link Cancellable} that have been previously added via {@link #add(Cancellable)} which have not yet
+     * been cancelled, and all future {@link Cancellable}s added via {@link #add(Cancellable)} will also be cancelled.
+     */
+    @Override
+    public void cancel() {
+        if (counterUpdater.getAndSet(this, Long.MIN_VALUE) >= 0) {
+            Throwable delayedCause = null;
+            for (Cancellable c : set) {
+                try {
+                    // Removal while iterating typically results in ConcurrentModificationException, but not for
+                    // ConcurrentHashMap. We use this approach to avoid concurrent invocation of cancel() between
+                    // this method and add (if they race).
+                    if (set.remove(c)) {
```

Comment:
IIRC it will be cheaper to create an `Iterator` and use `Iterator#remove()` here instead. We anyway cancel from a single thread, we don't need `CHM#remove(...)` guarantees here. WDYT?

Reply:
`remove()` doesn't let you know if you actually removed the element or not. this change is necessary if we are doing direct removal from `add` (as opposed to going through a guarded iterate all and remove method).